rsec16/coder.go

```diff
@@ -223,3 +223,38 @@ func (c Coder) ReconstructData(data, parity [][]byte) error {
 	}
 	return nil
 }
+
+// CanReconstructData tests wether or not enough information is present
+// to repair or not.
+func (c Coder) CanReconstructData(data, parity [][]byte) (bool, error) {
+	var availableRows, missingRows []int
+	var input [][]byte
+	for i, dataShard := range data {
+		if dataShard != nil {
+			availableRows = append(availableRows, i)
+			input = append(input, dataShard)
+		} else {
+			missingRows = append(missingRows, i)
+		}
+	}
+
+	if len(missingRows) == 0 {
+		// Nothing to reconstruct. TODO: what error is this?
```

Comment:
I did that first, but changed it to leaving the function signature intact and test for the specific error later on (in `main.go`), because that seemed to be the goal of the `errors` package (I read in a few SO posts you should simply test for named `error`s and since it was already returning a bool and err related to the status of the verification (as does nearly every function), I thought that adding in an extra enum does not help understand what's going on. I also don't think `Verify()` should be an exception, I like repair and create to also return the right state, so, the idiomatic thing to do I think is use named `error`s. And to be able to test for them requires an easily importable location, so there we are :)

I took `nil` to mean `OK` or `SUCCESS`, i.e. nothing to do. This is the practical meaning of `nil` elsewhere, right?

Would you object to keeping it like it is?

Reply:
Okay, I see what you mean -- `ReconstructData` would need to return an error for irrepairable error, so `CanReconstructData` may as well use the same interface. So that part I agree with -- `CanReconstructData` can return an `error` value, and if it's `nil`, either there's nothing to do or there's repairable damage. The caller can distinguish those cases by e.g. checking for missing data shards or some other way.

However, I don't think all the errors need to be converted to sentinel values -- for one, that would make all of them part of the public interface, and only a few of them need to be. So I'd prefer a single new error type for IrrepairableError, defined in `coder.go` near `CanReconstructData` and `ReconstructData`.

Also, I'm not a huge fan of sentinel error values, even though it's still a popular pattern. For one, the variables are public and thus mutatable outside of the package. Also, it precludes adding additional info as part of the error later on.

So a new error type as described in the beginning of https://blog.golang.org/go1.13-errors would be best, I think. Does that make sense?

Reply:
With the new error type you mean sth like this:
```
type NotFoundError struct {
    Name string
}

func (e *NotFoundError) Error() string { return e.Name + ": not found" }

if e, ok := err.(*NotFoundError); ok {
    // e.Name wasn't found
}
```

I encountered one in the codebase! And turned it into a sentinel `error` ;)

Onto your criticism:
> For one, the variables are public and thus mutatable outside of the package
That is not true for error types?
> Also, it precludes adding additional info as part of the error later on.
Sure, but that is already true for all errors, right? With gathering the errors in one file, I discovered that a few were signalling the same thing (not just identical error strings). I like collecting the sentinels in one place, because that's the only practical way of seeing similar/same errors (still something that could be done: reduce the list of sentinels by merging similar/same errors). But if you really dislike it, I can undo it and just make error types for the ones I actually need to catch right now.

Reply:
> With the new error type you mean sth like this:
> 
> I encountered one in the codebase! And turned it into a sentinel `error` ;)

Yes, something like that! I forgot I already had one in the code base -- you can copy that one, except it'd be a public type.

> > For one, the variables are public and thus mutatable outside of the package
> 
> That is not true for error types?

Correct, they're just types, so a new instance is created every time there's an error. The caller can modify them I suppose, but that's fine, since they're just local variables at that point. The sentinel values are mutable global variables, which are bad in general (and sloppy of the Go team to promote, IMO).

> > Also, it precludes adding additional info as part of the error later on.
> 
> Sure, but that is already true for all errors, right? With gathering the errors in one file, I discovered that a few were signalling the same thing (not just identical error strings). I like collecting the sentinels in one place, because that's the only practical way of seeing similar/same errors (still something that could be done: reduce the list of sentinels by merging similar/same errors). But if you really dislike it, I can undo it and just make error types for the ones I actually need to catch right now.

Yeah, that's correct for the status quo, but it's fine since that's (effectively) not public. So `errors.New` is the quick-and-dirty error solution; once a particular error is ready to "graduate", it can be promoted to a real (possibly public) error type possibly with parameters, etc.

Out of curiosity, which similar/same errors did you find? And yeah, I'd prefer you undo it for now. Sorry you went through all the trouble!

Reply:
I've updated the PR accordingly. The not great thing about defining the error in the file it is raised, is that now you have to now know not only about the error, but also the package/dir in which it is defined:
https://github.com/akalin/gopar/pull/4/files#diff-9b35b7853a837ea6f045f02b3d9039b133f98ca393b4b226930c6d476de9bf0dR448

Here's the error struct that you already had (you needed to test for that condition somewhere):
https://github.com/akalin/gopar/search?q=noPacketsFoundError

In a few places (tests mainly I think) you compare errors another way though, like `require.Equal(t, errors.New("invalid ASCII character"), err)` (where `err` is supposed to be that same error). Is that then also something that should be an error type?

My last question is then what to do with the commented out bit:
https://github.com/akalin/gopar/pull/4/files#diff-9ed8461bf845840bb7306d0b77157b768a58c0c9a3a7d4feb6c83903aa6dc651R537